nautobot/dcim/models/device_components.py

```diff
@@ -94,6 +98,59 @@ def to_objectchange(self, action, **kwargs):
         return super().to_objectchange(action, related_object=device, **kwargs)
 
 
+class ModularComponentModel(ComponentModel):
+    device = ForeignKeyWithAutoRelatedName(
+        to="dcim.Device",
+        on_delete=models.CASCADE,
+        blank=True,
+        null=True,
+    )
+    module = ForeignKeyWithAutoRelatedName(
+        to="dcim.Module",
+        on_delete=models.CASCADE,
+        blank=True,
+        null=True,
+    )
+
+    natural_key_field_names = ["name", "device", "module"]
+
+    class Meta:
+        abstract = True
+        constraints = [
+            models.CheckConstraint(
```

Comment:
Probably need a short comment here explaining what this does. Same for the other CheckConstraints added in this PR